packages/js/components/src/experimental-tree-control/hooks/use-keyboard.ts

```diff
@@ -0,0 +1,172 @@
+/**
+ * External dependencies
+ */
+
+/**
+ * Internal dependencies
+ */
+import { LinkedTree } from '../types';
+
+function getFirstChild(
+	currentHeading: HTMLDivElement
+): HTMLLabelElement | null {
+	const parentTreeItem = currentHeading?.closest< HTMLDivElement >(
+		'.experimental-woocommerce-tree-item'
+	);
+	const firstSubTreeItem = parentTreeItem?.querySelector(
+		'.experimental-woocommerce-tree > .experimental-woocommerce-tree-item'
+	);
+	const label = firstSubTreeItem?.querySelector< HTMLLabelElement >(
+		'.experimental-woocommerce-tree-item__heading > .experimental-woocommerce-tree-item__label'
+	);
+	return label ?? null;
+}
+
+function getFirstAncestor(
+	currentHeading: HTMLDivElement
+): HTMLLabelElement | null {
+	const parentTree = currentHeading?.closest< HTMLDivElement >(
+		'.experimental-woocommerce-tree'
+	);
+	const grandParentTreeItem = parentTree?.closest< HTMLDivElement >(
+		'.experimental-woocommerce-tree-item'
+	);
+	const label = grandParentTreeItem?.querySelector< HTMLLabelElement >(
+		'.experimental-woocommerce-tree-item__heading > .experimental-woocommerce-tree-item__label'
+	);
+	return label ?? null;
+}
+
+function getAllHeadings(
+	currentHeading: HTMLDivElement
+): NodeListOf< HTMLDivElement > | undefined {
+	const rootTree = currentHeading.closest< HTMLDivElement >(
+		'.experimental-woocommerce-tree--level-1'
```

Comment:
Is that `-1` here correct? Can't that number change depending on the situation?

Reply:
> Is that -1 here correct?

Yes it's. This `--level-1` refers to the nesting level of the tree. So that all subtrees have their parent level + 1 like `--level-2`, `--level-3` ... going from the root to the leafs.